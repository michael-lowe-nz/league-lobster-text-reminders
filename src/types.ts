export type Contacts = {
  Teams: Array<Team>;
};

export type Team = {
  Name: string;
  Players: Array<{ Name: string; Number: string }>;
};

export type Game = {
  dateString: string;
  time: string;
  court: string;
  teamAgainst: string;
};
