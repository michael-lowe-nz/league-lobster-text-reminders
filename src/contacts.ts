const examplePhoneNumber = "+18005550100";

export default {
  Test: {
    Teams: [
      {
        Name: "Los Angeles Lakers",
        Players: [
          {
            Name: "Shaquille ONeal",
            Number: process.env.TEST_PHONE_NUMBER || examplePhoneNumber,
          },
        ],
      },
    ],
  },
};
