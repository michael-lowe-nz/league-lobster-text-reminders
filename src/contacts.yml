---
teams:
  - Name: Gen-X
    Players:
      - Name: Michael
        Number: "+64272120047"
  - Name: "Craig's Team"
    Players:
      - Name: "Craig"
        Number: "+64272120047"
