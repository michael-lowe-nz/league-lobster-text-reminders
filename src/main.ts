import { App, Stack, StackProps } from "aws-cdk-lib";
import { OIDCSetup } from "./lib/stacks/oidcSetup";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Topic } from "aws-cdk-lib/aws-sns";
import { SmsSubscription } from "aws-cdk-lib/aws-sns-subscriptions";
import { Construct } from "constructs";
import { readFileSync } from "fs";
import { load } from "js-yaml";
import { Team } from "./types";

export class LeagueLobsterTextReminder extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const data = readFileSync("./src/contacts.yml", "utf-8");
    const contacts: any = load(data);

    contacts.teams.forEach((team: Team) => {
      const teamTopic = new Topic(this, `${team.Name}Alerts`);
      const teamAlertFunction = new NodejsFunction(
        this,
        `Alert${team.Name}Function`,
        {
          runtime: Runtime.NODEJS_18_X,
          entry: "./src/main.GetGamesFunction.ts",
          environment: {
            SNS_TOPIC_ARN: teamTopic.topicArn,
          },
        }
      );
      teamTopic.grantPublish(teamAlertFunction);
      team.Players.forEach((player: any) => {
        teamTopic.addSubscription(new SmsSubscription(player.Number));
      });
    });
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

/**
 * This is what happens when you run a local synth without specifying a NODE_ENV
 */
if (!process.env.NODE_ENV) {
  new LeagueLobsterTextReminder(app, "league-lobster-text-reminders-dev", {
    env: devEnv,
  });
}

if (process.env.NODE_ENV === "ci") {
  new OIDCSetup(app, "oidc-setup", { env: devEnv });
  new LeagueLobsterTextReminder(app, "league-lobster-text-reminders-prod", {
    env: devEnv,
  });
}

app.synth();
