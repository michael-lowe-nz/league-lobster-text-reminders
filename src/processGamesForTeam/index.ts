import { PublishCommand, SNSClient } from "@aws-sdk/client-sns";
import { Handler } from "aws-lambda";
import axios from "axios";
// import { getRandomEmoji } from "./lib/getRandomEmoji";
import moment from "moment";
import {
  extractGamesFromHTML,
  getNextGameAfterDate,
  getGameInfoFromGame,
} from "../lib/gamesFunctions";
import { Game } from "../types";

export const handler: Handler = async () => {
  const client = new SNSClient({ region: process.env.AWS_REGION });

  /**
   * @todo should be an environment variable
   */
  const url = `https://websites.mygameday.app/team_info.cgi?c=0-2854-0-622183-27083591&a=SFIX`;

  const requestResponse: any = await axios.get(url);
  const games: Array<Game> = extractGamesFromHTML(requestResponse.data);
  const nextGame: Game | null = getNextGameAfterDate(games, moment());

  if (!nextGame) {
    return {
      statusCode: 200,
      body: {
        message: "No game coming up this week",
      },
    };
  }

  const gameInfo: string = getGameInfoFromGame(nextGame);
  const Message = `üèÄ Gen-X Reminders üèÄ\nHey guys! Game is on ${gameInfo}`;
  const params = {
    Message,
    TopicArn: process.env.SNS_TOPIC_ARN,
    DefaultSMSType: "Promotional",
  };
  const response = await client.send(new PublishCommand(params));
  return {
    statusCode: 200,
    body: { snsResponse: response, message: Message },
  };
};
