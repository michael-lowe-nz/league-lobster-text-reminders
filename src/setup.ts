import { App } from "aws-cdk-lib";
import { OIDCSetup } from "./lib/stacks/oidcSetup";

const app = new App();

new OIDCSetup(app, "OIDCSetup", {
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION,
  },
});

app.synth();
