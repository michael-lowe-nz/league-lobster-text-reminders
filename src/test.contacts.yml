---
Teams:
  - Name: Gen-X
    Players:
      - Name: Michael
        Number: "+18005550100"
  - Name: "Craig's Team"
    Players:
      - Name: "Craig"
        Number: "+18005550100"
