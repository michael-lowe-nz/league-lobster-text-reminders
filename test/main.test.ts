import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { LeagueLobsterTextReminder } from "../src/main";
import { readFile } from "fs/promises";
import { load } from "js-yaml";
import { Team } from "../src/types";

test("Snapshot", () => {
  const app = new App();
  const stack = new LeagueLobsterTextReminder(app, "test");

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});

test("Test that the stack builds an SNS topic", () => {
  const app = new App();
  const stack = new LeagueLobsterTextReminder(app, "test");

  const template = Template.fromStack(stack);
  expect(template.hasResource("AWS::SNS::Topic", {}));
});

test("Test that the stack builds an SNS topic with the right number of subscribers in it based on the actual contacts.yml", async () => {
  const app = new App();
  const stack = new LeagueLobsterTextReminder(app, "test");

  const template = Template.fromStack(stack);
  const yaml = await readFile("./src/contacts.yml", "utf-8");
  const data: any = load(yaml);
  const numberOfContacts = data.teams.reduce(
    (totalContacts: number, team: Team) => {
      return totalContacts + team.Players.length;
    },
    0
  );
  expect(template.resourceCountIs("AWS::SNS::Subscription", numberOfContacts));
});

/**
 * I realised halfway through that a test like this would need a contacts list to be fed
 * into the stack as a prop. i.e. the stack always goes and creates its own thing
 */

// test("Test stack builds the right number of subscribers given a larger contacts list", () => {
//   const app = new App();
//   const stack = new LeagueLobsterTextReminder(app, "test");
//   const template = Template.fromStack(stack);

//   const Contacts = {
//     teams: [
//       {
//         Name: "Team 1",
//         Players: [
//           {
//             Name: "Player",
//             Number: "+6422",
//           },
//           {
//             Name: "Player2",
//             Number: "+6422",
//           },
//           {
//             Name: "Player3",
//             Number: "+6422",
//           },
//           {
//             Name: "Player4",
//             Number: "+6422",
//           },
//         ],
//       },
//       {
//         Name: "Team 2",
//         Players: [
//           {
//             Name: "Player",
//             Number: "+6422",
//           },
//           {
//             Name: "Player2",
//             Number: "+6422",
//           },
//           {
//             Name: "Player3",
//             Number: "+6422",
//           },
//           {
//             Name: "Player4",
//             Number: "+6422",
//           },
//         ],
//       },
//       {
//         Name: "Team 3",
//         Players: [
//           {
//             Name: "Player",
//             Number: "+6422",
//           },
//           {
//             Name: "Player2",
//             Number: "+6422",
//           },
//           {
//             Name: "Player3",
//             Number: "+6422",
//           },
//           {
//             Name: "Player4",
//             Number: "+6422",
//           },
//         ],
//       },
//     ],
//   };

//   expect(template.resourceCountIs("AWS::SNS::Subscription", 9));
// });
