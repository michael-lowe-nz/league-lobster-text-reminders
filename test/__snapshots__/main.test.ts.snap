// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlertTestTeam2Function6D2517FA": {
      "DependsOn": [
        "AlertTestTeam2FunctionServiceRoleDefaultPolicyF77B6850",
        "AlertTestTeam2FunctionServiceRoleEA0E27EA",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2a64b3343a06bfefa641409c1d8dced4de39295ff7b4630ee0cac338e3fe8175.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SNS_TOPIC_ARN": {
              "Ref": "TestTeam2Alerts2BE53530",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AlertTestTeam2FunctionServiceRoleEA0E27EA",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AlertTestTeam2FunctionServiceRoleDefaultPolicyF77B6850": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "TestTeam2Alerts2BE53530",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlertTestTeam2FunctionServiceRoleDefaultPolicyF77B6850",
        "Roles": [
          {
            "Ref": "AlertTestTeam2FunctionServiceRoleEA0E27EA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AlertTestTeam2FunctionServiceRoleEA0E27EA": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AlertTestTeamFunction3257EF79": {
      "DependsOn": [
        "AlertTestTeamFunctionServiceRoleDefaultPolicy918B5A1B",
        "AlertTestTeamFunctionServiceRole496082A0",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2a64b3343a06bfefa641409c1d8dced4de39295ff7b4630ee0cac338e3fe8175.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "SNS_TOPIC_ARN": {
              "Ref": "TestTeamAlerts58363266",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AlertTestTeamFunctionServiceRole496082A0",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AlertTestTeamFunctionServiceRole496082A0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "AlertTestTeamFunctionServiceRoleDefaultPolicy918B5A1B": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "TestTeamAlerts58363266",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AlertTestTeamFunctionServiceRoleDefaultPolicy918B5A1B",
        "Roles": [
          {
            "Ref": "AlertTestTeamFunctionServiceRole496082A0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestTeam2Alerts2BE53530": {
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "snskmskey1F6AC6F6",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
    "TestTeam2Alerts362D23E2E": {
      "Properties": {
        "Endpoint": "+3",
        "Protocol": "sms",
        "TopicArn": {
          "Ref": "TestTeam2Alerts2BE53530",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestTeam2Alerts4320BD087": {
      "Properties": {
        "Endpoint": "+4",
        "Protocol": "sms",
        "TopicArn": {
          "Ref": "TestTeam2Alerts2BE53530",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestTeamAlerts17FF09393": {
      "Properties": {
        "Endpoint": "+1",
        "Protocol": "sms",
        "TopicArn": {
          "Ref": "TestTeamAlerts58363266",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestTeamAlerts2E7EDE869": {
      "Properties": {
        "Endpoint": "+2",
        "Protocol": "sms",
        "TopicArn": {
          "Ref": "TestTeamAlerts58363266",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "TestTeamAlerts58363266": {
      "Properties": {
        "KmsMasterKeyId": {
          "Fn::GetAtt": [
            "snskmskey1F6AC6F6",
            "Arn",
          ],
        },
      },
      "Type": "AWS::SNS::Topic",
    },
    "snskmskey1F6AC6F6": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Description": "KMS key for encrypting the objects in an S3 bucket",
        "EnableKeyRotation": false,
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PendingWindowInDays": 7,
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Delete",
    },
    "snskmskeyAlias7A51B3E5": {
      "Properties": {
        "AliasName": "alias/sns/league-lobster-key",
        "TargetKeyId": {
          "Fn::GetAtt": [
            "snskmskey1F6AC6F6",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
